R/tar_quarto_files.R

```diff
@@ -51,15 +51,18 @@ tar_quarto_files <- function(path = ".", profile = NULL) {
     tar_quarto_files_document(path)
   )
   for (field in c("sources", "output", "input")) {
-    out[[field]] <- sort(fs::path_rel(unique(as.character(out[[field]]))))
-    out[[field]] <- as.character(out[[field]])
+    out[[field]] <- as.character(fs::path_rel(out[[field]]))
+    out[[field]] <- unique(sort(out[[field]]))
   }
   out
 }
 
 tar_quarto_files_document <- function(path) {
   info <- quarto::quarto_inspect(input = path)
-  out <- list(sources = path)
+  out <- list()
+
+  # Collect data about source files.
+  out$sources <- tar_quarto_files_get_source_files(info$fileInformation)
```

Comment:
As currently written, this function is based on detecting calls to `tar_read()` and `tar_load()`. `tar_render()` source files do not necessarily call these functions. Is there a more reliable way to detect source files? If we know they're all qmd files that will be run/rendered, can we just include all of them?

Reply:
I'm not sure whether I do understand your comment correctly. Is your question whether files that do not use `tar_read()` or `tar_load()` are still somehow detected and used as a dependency?

The doc says the following:

> A named list of important file paths in a Quarto project or document:
> - `sources`: source files with [tar_load()](https://docs.ropensci.org/targets/reference/tar_load.html)/[tar_read()](https://docs.ropensci.org/targets/reference/tar_read.html) target dependencies in R code chunks.
> - `output`: output files that will be generated during [quarto::quarto_render()](https://quarto-dev.github.io/quarto-r/reference/quarto_render.html).
> - `input`: pre-existing files required to render the project or document, such as _quarto.yml.

I followed this documentation and collected only files that have `tar_read` or a `tar_load` in them in `sources`. All other included files that do not call `tar_read/load` are in `input`. See [this test](https://github.com/ropensci/tarchetypes/pull/200/files/d6f1772478521b35cd89ec036aee013f1940979b#diff-46560dd7d232f074dec1eed5dc1592f0bbaf6862b4a66abd362c2867b8072c75R177).

To be honest, I'm not sure whether the differentiation between these three list items is necessary for targets as all of them are simply treated as file dependencies (at least this way I understand the code). In addition, it would be really easy for `tar_quarto_files` to detect which targets are loaded because `fileInformation` contains the code cells. This way, the call to `knitr_deps` in `tar_quarto_raw` could be eliminated. But I didn't change that to not alter the interface of these functions.

Did I understand your question correctly?

Reply:
If your question is whether also files are detected that are included by the traditional knitr (`tar_render`) way (` ```{r child="file1"`): this is the case.

Reply:
> I followed this documentation and collected only files that have tar_read or a tar_load in them in sources. 

Sources *can* have `tar_read()`/`tar_load()`, but they don't have to, and they don't need to be distinguished from input files that don't have `tar_read()`/`tar_load()`. `tarchetypes` just needs to know which files to [*scan* for `tar_read()`/`tar_load()` using static code analysis](https://github.com/ropensci/tarchetypes/blob/b0fb6f2f254551d9fd2c7fc03a7092540a0cf76e/R/tar_quarto_raw.R#L158), and it needs to know the inputs to `quarto_render()`.

> To be honest, I'm not sure whether the differentiation between these three list items is necessary for targets as all of them are simply treated as file dependencies (at least this way I understand the code). 

The sources are the files scanned with `knitr_deps()`. The other input files do not necessarily even have code chunks that can be `parse()`'d as valid R code.

> In addition, it would be really easy for tar_quarto_files to detect which targets are loaded because fileInformation contains the code cells.

That part could be useful for just extracting the code chunks, but we already can use `knits::knit(tangle = TRUE)`, and it is a unified approach that also works with R Markdown and `knitr`.

> This way, the call to knitr_deps in tar_quarto_raw could be eliminated. 

Regardless of where we get the text of the code chunks, `knitr_deps()`  is still needed because it runs static code analysis to robustly detect `tar_load()`/`tar_read()` calls and the targets that they load/read. It is much more reliable to [walk the AST](https://adv-r.hadley.nz/expressions.html#ast-funs) than manipulate text.

Reply:
Thank you for your feedback and the clarifications!

> Sources can have tar_read()/tar_load(), but they don't have to, and they don't need to be distinguished from source files that don't have tar_read()/tar_load(). tarchetypes just needs to know which files to [scan for tar_read()/tar_load() using static code analysis](https://github.com/ropensci/tarchetypes/blob/b0fb6f2f254551d9fd2c7fc03a7092540a0cf76e/R/tar_quarto_raw.R#L158), and it needs to know the inputs to quarto_render().

Thanks for clarifying this. I did not understand the documentation in such a way. I changed the documentation to make this point clearer. Additionally, I reworked the code and the tests to reflect this. Things are much simpler now.

Regarding `knitr_deps()`: Sorry for not being clear here. My intention was not to completely remove the logic behind `knitr_deps()` but to not read the files once again as `quarto::quarto_inspect` has already done this. For example, one could create an AST from the code snippets provided by quarto inspect. However, in all my reports, report generation takes much longer than tarchetypes to decide whether to rerun the report or not. Thus, it probably doesn't matter and it is the easiest solution to keep everything as is.